From ea064317c49b86e2f47ef4db48a0df847279ee1f Mon Sep 17 00:00:00 2001
From: Andre Heinecke <aheinecke@intevation.de>
Date: Wed, 18 May 2011 19:42:47 +0200
Subject: [PATCH] Add FullShutdown button for windows.

    On Windows when Kontact is the only KDE application used
    it is necessary to provide the user with a GUI option to
    completely shut down everything that was started by Kontact.

    This patch adds another button in Kontacts menu to do this.
---
 kontact/src/kontactui.rc   |    1 +
 kontact/src/mainwindow.cpp |   26 +++++++++++++++++++++++++-
 kontact/src/mainwindow.h   |    1 +
 3 files changed, 27 insertions(+), 1 deletions(-)

diff --git a/kontact/src/kontactui.rc b/kontact/src/kontactui.rc
index 9125e14..f35536e 100644
--- a/kontact/src/kontactui.rc
+++ b/kontact/src/kontactui.rc
@@ -8,6 +8,7 @@
     <Separator/>
     <Action name="action_new"/>
     <Separator/>
+    <Action name="full_shutdown"/>
     <Action name="file_quit"/>
     <Merge/>
   </Menu>
diff --git a/kontact/src/mainwindow.cpp b/kontact/src/mainwindow.cpp
index 09053ab..cbec31d 100644
--- a/kontact/src/mainwindow.cpp
+++ b/kontact/src/mainwindow.cpp
@@ -33,6 +33,7 @@ using namespace Kontact;
 #include <libkdepim/broadcaststatus.h>
 #include <libkdepim/progressdialog.h>
 #include <libkdepim/statusbarprogresswidget.h>
+#include <akonadi/servermanager.h>
 
 #include <KontactInterface/Core>
 #include <KontactInterface/Plugin>
@@ -68,6 +69,7 @@ using namespace Kontact;
 #include <KSycoca>
 
 #include <QDBusConnection>
+#include <QProcess>
 #include <QSplitter>
 #include <QStackedWidget>
 #include <QTimer>
@@ -414,6 +416,7 @@ void MainWindow::initAboutScreen()
 
 void MainWindow::setupActions()
 {
+
   actionCollection()->addAction( KStandardAction::Quit, this, SLOT(slotQuit()) );
 
   mNewActions = new KActionMenu(
@@ -423,6 +426,10 @@ void MainWindow::setupActions()
   mNewActions->setShortcut( KStandardShortcut::openNew() );
   connect( mNewActions, SIGNAL(triggered(bool)), this, SLOT(slotNewClicked()) );
 
+  KAction *action = new KAction(KIcon( "system-shutdown" ),  i18n( "Full Shutdown" ), this );
+  actionCollection()->addAction( "full_shutdown" , action );
+  connect( action, SIGNAL( triggered() ), SLOT( slotFullShutdown() ) );
+
   // If the user is using disconnected imap mail folders as groupware, we add
   // plugins' Synchronize actions to the toolbar which trigger an imap sync.
   // Otherwise it's redundant and misleading.
@@ -446,7 +453,7 @@ void MainWindow::setupActions()
     connect( mSyncActions, SIGNAL(triggered(bool)), this, SLOT(slotSyncClicked()) );
   }
 
-  KAction *action =
+  action =
     new KAction( KIcon( "configure" ),
                  i18nc( "@action:inmenu", "Configure Kontact..." ), this );
   action->setHelpText(
@@ -998,6 +1005,23 @@ void MainWindow::showTip( bool force )
   KTipDialog::showMultiTip( this, tips, force );
 }
 
+void MainWindow::slotFullShutdown()
+{
+  const QString message = i18n( "A full shutdown will disable notifications\nabout new emails and upcoming events." );
+  const int result = KMessageBox::warningContinueCancel( 0, message );
+
+  if ( result == KMessageBox::Cancel )
+    return;
+
+  Akonadi::ServerManager::self()->stop(); 
+#ifdef Q_OS_WIN
+  const QString path = KStandardDirs::findExe( QLatin1String("kdeinit4" ) );
+  QProcess::startDetached( path + " --terminate" );
+#endif
+  // TODO: Kill KOrgac Nepomuk and so on softly
+  close();
+}
+
 void MainWindow::slotQuit()
 {
   mReallyClose = true;
diff --git a/kontact/src/mainwindow.h b/kontact/src/mainwindow.h
index d99b3e1..d9ddd8f 100644
--- a/kontact/src/mainwindow.h
+++ b/kontact/src/mainwindow.h
@@ -83,6 +83,7 @@ class KONTACT_EXPORT MainWindow : public KontactInterface::Core
     void slotNewClicked();
     void slotSyncClicked();
     void slotQuit();
+    void slotFullShutdown();
     void slotShowTip();
     void slotShowTipOnStart();
     void slotNewToolbarConfig();
-- 
1.7.4.4

